File: 2019/day05_1.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from typing import Any


def solution(content: str) -> Any:
    program = [int(x) for x in content.split(",")]
    current = 0
    halt = False
    output = None
    while not halt:
        current, program, halt, n_output = execute(current, program)
        output = n_output if n_output is not None else output
    return output


def execute(current: int, program: list[int]):
    current_digits = str(program[current]).rjust(5, "0")
    fst_mod = current_digits[2]
    snd_mod = current_digits[1]
    opcode = current_digits[3] + current_digits[4]

    if opcode not in ["01", "02", "03", "04", "99"]:
        raise ValueError(f"Invalid opcode {opcode} at {current} in {program}")

    copy = program.copy()
    if opcode == "01":
        copy[copy[current + 3]] = get_value(program, current, 1, fst_mod) + get_value(
            program, current, 2, snd_mod
        )
        return (current + 4, copy, False, None)
    if opcode == "02":
        copy[copy[current + 3]] = get_value(program, current, 1, fst_mod) * get_value(
            program, current, 2, snd_mod
        )
        return (current + 4, copy, False, None)
    if opcode == "03":
        copy[copy[current + 1]] = 1
        return (current + 2, copy, False, None)
    if opcode == "04":
        copy[copy[current + 1]] = get_value(program, current, 1, fst_mod)
        return (current + 2, copy, False, copy[copy[current + 1]])

    # if opcode == "99":
    return (current, copy, True, None)


def get_value(program: list[int], current: int, param: int, mode: str):
    if mode == "1":
        return program[current + param]
    return program[program[current + param]]


def main() -> None:
    with open("day05.dat", "r", encoding="utf8") as file:
        print(solution(file.read().strip()))


if __name__ == "__main__":
    main()

File: 2019/day05-1.ts
type Output = {
  index: number;
  program: number[];
  halt: boolean;
  output: number | null;
};

function solution(content: string) {
  return execute(content.split(",").map(Number));
}

function execute(program: number[]) {
  let copy = program.slice();
  let halt = false;
  let current = 0;
  let output = null;
  while (!halt) {
    const result = executeOne(current, copy);
    current = result.index;
    copy = result.program;
    halt = result.halt;
    output = result.output ?? output;
  }
  return output;
}

function executeOne(current: number, program: number[]): Output {
  const currentDigits = String(program[current]).padStart(5, "0");
  const opcode = Number(currentDigits[3] + currentDigits[4]);
  const fstMod = currentDigits[2];
  const sndMod = currentDigits[1];
  const copy = program.slice();

  if (opcode == 1) {
    copy[copy[current + 3]] =
      getValue(program, current, 1, fstMod) +
      getValue(program, current, 2, sndMod);
    return {
      index: current + 4,
      program: copy,
      halt: false,
      output: null,
    };
  }
  if (opcode == 2) {
    copy[copy[current + 3]] =
      getValue(program, current, 1, fstMod) *
      getValue(program, current, 2, sndMod);
    return {
      index: current + 4,
      program: copy,
      halt: false,
      output: null,
    };
  }
  if (opcode == 3) {
    copy[copy[current + 1]] = 1;
    return {
      index: current + 2,
      program: copy,
      halt: false,
      output: null,
    };
  }
  if (opcode == 4) {
    copy[copy[current + 1]] = getValue(program, current, 1, fstMod);
    return {
      index: current + 2,
      program: copy,
      halt: false,
      output: copy[copy[current + 1]],
    };
  }
  if (opcode == 99) {
    return {
      index: 0,
      program: [],
      halt: true,
      output: null,
    };
  }
  throw Error(
    `Invalid opcode ${opcode} at ${current} in ${program.slice(
      current,
      current + 4
    )}`
  );
}

function getValue(
  program: number[],
  current: number,
  param: number,
  mode: string
) {
  if (mode === "1") return program[current + param];
  return program[program[current + param]];
}

Deno.readTextFile("day05.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);
